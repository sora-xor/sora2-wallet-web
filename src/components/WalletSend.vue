<template>
  <wallet-base
    :title="t(`walletSend.${step === 1 ? 'title' : 'confirmTitle'}`)"
    :tooltip="tooltipContent"
    show-back
    :showHeader="showAdditionalInfo"
    @back="handleBack"
  >
    <div class="wallet-send">
      <template v-if="step === 1">
        <s-input
          class="wallet-send-address"
          :maxlength="128"
          :placeholder="t('walletSend.address')"
          v-model="address"
        />
        <template v-if="validAddress && isNotSoraAddress">
          <p class="wallet-send-address-warning">{{ t('walletSend.addressWarning') }}</p>
          <s-tooltip :content="copyTooltip">
            <p class="wallet-send-address-formatted" @click="handleCopyAddress(formattedSoraAddress)">
              {{ formattedSoraAddress }}
            </p>
          </s-tooltip>
        </template>
        <p v-if="isAccountAddress" class="wallet-send-address-error">{{ t('walletSend.addressError') }}</p>
        <s-float-input
          v-model="amount"
          class="wallet-send-input"
          size="medium"
          has-locale-string
          :delimiters="delimiters"
          :max="getMax((asset || {}).address)"
        >
          <div class="wallet-send-amount" slot="top">
            <div class="wallet-send-amount-title">{{ t('walletSend.amount') }}</div>
            <div class="wallet-send-amount-balance">
              <span class="wallet-send-amount-balance-title">{{ t('walletSend.balance') }}</span>
              <formatted-amount-with-fiat-value
                value-can-be-hidden
                fiat-format-as-value
                with-left-shift
                value-class="wallet-send-amount-balance-value"
                :value="balance"
                :asset-symbol="asset.symbol"
                :fiat-value="getFiatBalance(asset)"
              />
            </div>
          </div>
          <div class="asset s-flex" slot="right">
            <s-button
              v-if="isMaxButtonAvailable"
              class="asset-max s-typography-button--small"
              type="primary"
              alternative
              size="mini"
              border-radius="mini"
              @click="handleMaxClick"
            >
              {{ t('walletSend.max') }}
            </s-button>
            <div class="asset-box">
              <i class="asset-logo" :style="getAssetIconStyles(asset.address)" />
              <span class="asset-name">{{ asset.symbol }}</span>
            </div>
          </div>
          <div class="asset-info" slot="bottom">
            <formatted-amount v-if="fiatAmount" :value="fiatAmount" is-fiat-value />
            <div class="asset-highlight">
              {{ asset.name || asset.symbol }}
              <s-tooltip :content="copyTooltip">
                <span class="asset-id" @click="handleCopyAddress(asset.address)">
                  ({{ getFormattedAddress(asset) }})
                </span>
              </s-tooltip>
            </div>
          </div>
        </s-float-input>
        <s-button
          class="wallet-send-action s-typography-button--large"
          type="primary"
          :disabled="sendButtonDisabled"
          @click="handleSend"
        >
          {{ sendButtonDisabledText || t('walletSend.title') }}
        </s-button>
      </template>
      <template v-else-if="step === 2">
        <network-fee-warning :fee="formattedFee" @confirm="confirmNextTxFailure" />
      </template>
      <template v-else>
        <div class="confirm">
          <div class="confirm-asset s-flex">
            <span class="confirm-asset-title">{{ formatStringValue(amount, asset.decimals) }}</span>
            <div class="confirm-asset-value s-flex">
              <i class="asset-logo" :style="getAssetIconStyles(asset.address)" />
              <span class="asset-name">{{ asset.symbol }}</span>
            </div>
          </div>
          <div class="confirm-from">{{ account.address }}</div>
          <s-icon name="arrows-arrow-bottom-24" />
          <div class="confirm-to">{{ formattedSoraAddress }}</div>
        </div>
        <s-button
          class="wallet-send-action s-typography-button--large"
          type="primary"
          :disabled="sendButtonDisabled"
          @click="handleConfirm"
        >
          {{ sendButtonDisabledText || t('walletSend.confirm') }}
        </s-button>
      </template>
      <wallet-fee v-if="showAdditionalInfo" :value="fee" />
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';
import { Action, Getter } from 'vuex-class';
import {
  AccountAsset,
  Asset,
  FPNumber,
  CodecString,
  KnownAssets,
  KnownSymbols,
  Operation,
  XOR,
} from '@sora-substrate/util';

import TransactionMixin from './mixins/TransactionMixin';
import FormattedAmountMixin from './mixins/FormattedAmountMixin';
import NetworkFeeWarningMixin from './mixins/NetworkFeeWarningMixin';
import CopyAddressMixin from './mixins/CopyAddressMixin';
import { RouteNames } from '../consts';
import { formatAddress, formatSoraAddress, getAssetIconStyles } from '../util';
import { api } from '../api';

import WalletBase from './WalletBase.vue';
import FormattedAmount from './FormattedAmount.vue';
import FormattedAmountWithFiatValue from './FormattedAmountWithFiatValue.vue';
import NetworkFeeWarning from './NetworkFeeWarning.vue';
import WalletFee from './WalletFee.vue';

@Component({
  components: {
    WalletBase,
    FormattedAmount,
    FormattedAmountWithFiatValue,
    NetworkFeeWarning,
    WalletFee,
  },
})
export default class WalletSend extends Mixins(
  TransactionMixin,
  FormattedAmountMixin,
  CopyAddressMixin,
  NetworkFeeWarningMixin
) {
  readonly delimiters = FPNumber.DELIMITERS_CONFIG;

  @Getter currentRouteParams!: any;
  @Getter accountAssets!: Array<AccountAsset>;

  @Action navigate!: (options: { name: string; params?: object }) => Promise<void>;
  @Action transfer!: (options: { to: string; amount: string }) => Promise<void>;

  step = 1;
  address = '';
  amount = '';
  showWarningFeeNotification = false;
  showAdditionalInfo = true;

  get fee(): FPNumber {
    return this.getFPNumberFromCodec(this.networkFees.Transfer);
  }

  get formattedFee(): string {
    return this.fee.toLocaleString();
  }

  get tooltipContent(): string {
    return this.step === 1 ? this.t('walletSend.tooltip') : '';
  }

  get asset(): AccountAsset {
    const { address } = this.currentRouteParams.asset;

    return this.accountAssets.find((asset) => asset.address === address) || this.currentRouteParams.asset;
  }

  get balance(): string {
    return this.formatCodecNumber(this.asset.balance.transferable, this.asset.decimals);
  }

  get assetFiatPrice(): Nullable<CodecString> {
    return this.getAssetFiatPrice(this.asset);
  }

  get fiatAmount(): Nullable<string> {
    return this.getFiatAmountByString(this.amount, this.asset);
  }

  get emptyAddress(): boolean {
    if (!this.address.trim()) {
      return true;
    }
    return false;
  }

  get isAccountAddress(): boolean {
    return [this.address, this.formattedSoraAddress].includes(this.account.address);
  }

  get validAddress(): boolean {
    if (this.emptyAddress) {
      return false;
    }
    return api.validateAddress(this.address) && !this.isAccountAddress;
  }

  get formattedSoraAddress(): string {
    if (this.emptyAddress) {
      return '';
    }
    try {
      return formatSoraAddress(this.address);
    } catch {
      return '';
    }
  }

  get isNotSoraAddress(): boolean {
    return !!this.formattedSoraAddress && this.address.slice(0, 2) !== 'cn';
  }

  get emptyAmount(): boolean {
    return +this.amount === 0;
  }

  get validAmount(): boolean {
    const amount = this.getFPNumber(this.amount, this.asset.decimals);
    const balance = this.getFPNumberFromCodec(this.asset.balance.transferable, this.asset.decimals);
    return amount.isFinity() && !amount.isZero() && FPNumber.lte(amount, balance);
  }

  get isMaxButtonAvailable(): boolean {
    const decimals = this.asset.decimals;
    const balance = this.getFPNumberFromCodec(this.asset.balance.transferable, decimals);
    const amount = this.getFPNumber(this.amount, decimals);
    if (this.isXorAccountAsset(this.asset)) {
      if (this.fee.isZero()) {
        return false;
      }
      return !FPNumber.eq(this.fee, balance.sub(amount)) && FPNumber.gt(balance, this.fee);
    }
    return !FPNumber.eq(balance, amount);
  }

  get hasEnoughXor(): boolean {
    return api.hasEnoughXor(this.asset, this.amount, this.fee);
  }

  get sendButtonDisabled(): boolean {
    return this.loading || !this.validAddress || !this.validAmount || !this.hasEnoughXor;
  }

  get sendButtonDisabledText(): string {
    if (!this.validAddress) {
      return this.t(`walletSend.${this.emptyAddress ? 'enterAddress' : 'badAddress'}`);
    }

    if (!this.validAmount) {
      return this.t(
        `walletSend.${this.emptyAmount ? 'enterAmount' : 'badAmount'}`,
        this.emptyAmount ? {} : { tokenSymbol: this.asset.symbol }
      );
    }

    if (!this.hasEnoughXor) {
      return this.t('walletSend.badAmount', { tokenSymbol: KnownSymbols.XOR });
    }

    return '';
  }

  get xorBalance(): Nullable<CodecString> {
    // TODO: XOR balance here can be unsyncronized, need to fix
    const accountXor = api.accountAssets.find((asset) => asset.address === XOR.address);
    return accountXor ? accountXor.balance.transferable : null;
  }

  isXorAccountAsset(asset: AccountAsset): boolean {
    const knownAsset = KnownAssets.get(asset.address);
    if (!knownAsset) {
      return false;
    }
    return knownAsset.symbol === KnownSymbols.XOR;
  }

  getAssetIconStyles = getAssetIconStyles;

  getFormattedAddress(asset: AccountAsset): string {
    return formatAddress(asset.address, 10);
  }

  handleBack(): void {
    if (this.step !== 1) {
      this.showAdditionalInfo = true;
      this.step = 1;
      return;
    }
    this.navigate({ name: RouteNames.Wallet });
  }

  async handleMaxClick(): Promise<void> {
    if (this.isXorAccountAsset(this.asset)) {
      const balance = this.getFPNumberFromCodec(this.asset.balance.transferable, this.asset.decimals);
      this.amount = balance.sub(this.fee).toString();
      return;
    }
    this.amount = this.getStringFromCodec(this.asset.balance.transferable, this.asset.decimals);
  }

  async handleSend(): Promise<void> {
    if (
      !this.isXorSufficientForNextTx({
        type: Operation.Transfer,
        isXorAccountAsset: this.isXorAccountAsset(this.asset),
        xorBalance: this.xorBalance ? this.getFPNumberFromCodec(this.xorBalance) : this.Zero,
        amount: this.getFPNumber(this.amount),
      })
    ) {
      this.showAdditionalInfo = false;
      this.step = 2;
      return;
    }
    this.step = 3;
  }

  async handleConfirm(): Promise<void> {
    await this.withNotifications(async () => {
      if (!this.hasEnoughXor) {
        throw new Error('walletSend.badAmount');
      }
      await this.transfer({ to: this.address, amount: this.amount });
    });
    this.navigate({ name: RouteNames.Wallet });
  }

  confirmNextTxFailure(): void {
    this.showAdditionalInfo = true;
    this.step = 3;
  }
}
</script>

<style lang="scss">
.wallet-send {
  &-amount-balance {
    .formatted-amount--fiat-value {
      text-align: right;
    }
  }
  &-input .el-input__inner {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    font-size: var(--s-font-size-large);
    line-height: var(--s-line-height-small);
    font-weight: 800;
  }
}
</style>

<style scoped lang="scss">
@import '../styles/icons';
$logo-size: var(--s-size-mini);
// TODO: fix typography issues here
.wallet-send {
  .asset {
    align-items: center;

    & > * {
      margin-left: var(--s-basic-spacing);
    }

    &-balance {
      margin-left: auto;
      &-title {
        color: var(--s-color-base-content-tertiary);
        font-size: var(--s-font-size-mini);
      }
      &-value {
        margin-left: $basic-spacing-mini;
        letter-spacing: var(--s-letter-spacing-big);
      }
    }
    &-box {
      display: flex;
      align-items: center;
      background-color: var(--s-color-utility-surface);
      border-radius: var(--s-border-radius-mini);
      box-shadow: var(--s-shadow-element);
      padding: $basic-spacing-mini #{$basic-spacing-extra-small};
    }
    &-logo {
      @include asset-logo-styles(var(--s-size-mini));
      margin-right: var(--s-basic-spacing);
    }
    &-max {
      height: var(--s-size-mini);
      padding: $basic-spacing-mini var(--s-basic-spacing);
    }
    &-max,
    &-name {
      font-weight: 800;
    }
    &-name {
      font-size: var(--s-icon-font-size-small);
      line-height: var(--s-line-height-reset);
    }
    &-info {
      display: flex;
      align-items: baseline;
      .formatted-amount--fiat-value {
        margin-right: $basic-spacing-mini;
        font-weight: 600;
      }
    }
    &-highlight {
      margin-left: auto;
      color: var(--s-color-base-content-secondary);
      font-size: var(--s-font-size-extra-mini);
      font-weight: 300;
      line-height: var(--s-line-height-medium);
      letter-spacing: var(--s-letter-spacing-small);
      text-align: right;
    }
  }
  .asset-id,
  &-address-formatted {
    cursor: pointer;
    &:hover {
      text-decoration: underline;
    }
  }
  &-address-formatted,
  .confirm-from,
  .confirm-to {
    word-break: break-all;
  }
  &-address {
    margin-bottom: var(--s-basic-spacing);
    &-description {
      margin-bottom: #{$basic-spacing-medium};
    }
    &-warning,
    &-error,
    &-formatted {
      padding-right: calc(var(--s-basic-spacing) * 1.25);
      padding-left: calc(var(--s-basic-spacing) * 1.25);
    }
    &-warning,
    &-error {
      margin-bottom: var(--s-basic-spacing);
      font-weight: 400;
      font-size: var(--s-font-size-extra-small);
      line-height: var(--s-line-height-base);
    }
    &-warning {
      color: var(--s-color-status-warning);
    }
    &-error {
      color: var(--s-color-status-error);
    }
    &-formatted {
      margin-bottom: calc(var(--s-basic-spacing) * 2);
      font-weight: 200;
      font-size: var(--s-font-size-mini);
      line-height: var(--s-line-height-base);
      letter-spacing: var(--s-letter-spacing-small);
    }
  }
  &-amount {
    display: flex;
    justify-content: space-between;
    align-items: baseline;
    margin-bottom: var(--s-basic-spacing);
    font-size: var(--s-font-size-mini);
    font-weight: 300;
    line-height: var(--s-line-height-medium);
    text-transform: uppercase;

    &-title,
    &-balance {
      display: inline-flex;
      align-items: baseline;

      &-title {
        margin-right: $basic-spacing-mini;
      }
    }
    &-balance {
      flex-wrap: wrap;
      justify-content: flex-end;
      margin-left: var(--s-basic-spacing);
      &-title {
        color: var(--s-color-base-content-secondary);
      }
    }
  }
  &-action {
    margin-top: #{$basic-spacing-medium};
    width: 100%;
  }
  .confirm {
    &-asset {
      margin-bottom: #{$basic-spacing-medium};
      font-size: var(--s-heading2-font-size);
      line-height: var(--s-line-height-small);
      font-weight: 800;
      &-title {
        line-height: 1.33;
        flex: 1;
        word-break: break-word;
      }
      &-value {
        align-items: center;
        justify-content: flex-end;
        white-space: nowrap;
        .asset {
          &-logo {
            @include asset-logo-styles(var(--s-size-small));
            margin-right: calc(var(--s-basic-spacing) * 2);
          }
          &-name {
            font-size: var(--s-heading2-font-size);
            line-height: var(--s-line-height-small);
          }
        }
      }
    }
    &-from {
      margin-bottom: var(--s-basic-spacing);
    }
    &-to {
      margin-top: var(--s-basic-spacing);
      overflow-wrap: break-word;
    }
    &-from,
    &-to {
      // It's set to small size cuz we need to show full address
      font-size: var(--s-font-size-mini);
      font-weight: 600;
    }
  }
}
</style>
