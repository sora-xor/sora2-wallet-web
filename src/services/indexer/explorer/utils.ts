import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
