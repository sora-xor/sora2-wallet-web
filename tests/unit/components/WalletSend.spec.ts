import Vuex from 'vuex'
import { shallowMount } from '@vue/test-utils'

import { useDescribe, localVue } from '../../utils'
import { MOCK_ACCOUNT_ASSETS, MOCK_WHITE_LIST, MOCK_ACCOUNTS, MOCK_NETWORK_FEE, MOCK_FIAT_PRICE_AND_APY_OBJECT } from '../../utils/mock'
import { MOCK_WALLET_SEND } from '../../utils/WalletSendMock'

import WalletSend from '@/components/WalletSend.vue'

useDescribe('WalletSend.vue', WalletSend, () => {
  const store = new Vuex.Store({
    getters: {
      currentRouteParams: () => ({ id: '1', asset: MOCK_ACCOUNT_ASSETS[0] }),
      accountAssets: () => MOCK_ACCOUNT_ASSETS,
      account: () => MOCK_ACCOUNTS[0],
      whitelist: () => MOCK_WHITE_LIST,
      networkFees: () => MOCK_NETWORK_FEE,
      fiatPriceAndApyObject: () => MOCK_FIAT_PRICE_AND_APY_OBJECT
    },
    actions: {
      setAttribute: jest.fn()
    }
  })
  MOCK_WALLET_SEND.map(item => it(`[${item.title}]: should be rendered correctly`, () => {
    const wrapper = shallowMount(WalletSend, {
      localVue,
      store,
      data () {
        return {
          step: item.step,
          address: item.address,
          amount: item.amount
        }
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  }))
})
