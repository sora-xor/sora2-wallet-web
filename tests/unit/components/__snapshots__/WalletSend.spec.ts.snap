// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletSend.vue [Confirm Step]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="2"
  showback="true"
  showheader="true"
  title="Confirm transaction"
  tooltip=""
>
  <div
    class="wallet-send"
  >
    <network-fee-warning-stub
      fee="0"
      payoff="true"
      symbol="XOR"
    />
     
    <wallet-fee-stub
      value="0"
    />
  </div>
</wallet-base-stub>
`;

exports[`WalletSend.vue [Create Step: Empty Fields]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="1"
  showback="true"
  showheader="true"
  title="Send"
  tooltip="Send tokens between SORA network accounts"
>
  <div
    class="wallet-send"
  >
    <address-book-input-stub
      class="wallet-send-address"
      excludeconnected="true"
      value=""
    />
     
    <!---->
     
    <!---->
     
    <s-float-input-stub
      class="wallet-send-input"
      decimals="18"
      delimiters="[object Object]"
      haslocalestring="true"
      max="100000000000000000000"
      size="medium"
      value=""
    >
      <div
        class="wallet-send-amount"
      >
        <div
          class="wallet-send-amount-title"
        >
          Amount
        </div>
         
        <div
          class="wallet-send-amount-balance"
        >
          <span
            class="wallet-send-amount-balance-title"
          >
            Balance
          </span>
           
          <formatted-amount-with-fiat-value-stub
            assetsymbol="XOR"
            fiatfontsizerate="normal"
            fiatfontweightrate="normal"
            fiatformatasvalue="true"
            fiatvalue="151.782"
            fontsizerate="normal"
            fontweightrate="normal"
            hasfiatvalue="true"
            value="123.4"
            valuecanbehidden="true"
            valueclass="wallet-send-amount-balance-value"
            withleftshift="true"
          />
        </div>
      </div>
       
      <div
        class="asset s-flex"
      >
        <!---->
         
        <div
          class="asset-box"
        >
          <div
            class="asset-box__logo"
          >
            <token-logo-stub
              size="small"
              token="[object Object]"
              tokensymbol=""
            />
          </div>
           
          <span
            class="asset-name"
          >
            XOR
          </span>
        </div>
      </div>
       
      <div
        class="asset-info"
      >
        <formatted-amount-stub
          assetsymbol=""
          fiatsign="$"
          fontsizerate="normal"
          fontweightrate="normal"
          isfiatvalue="true"
          value="0"
        />
         
        <div
          class="asset-highlight"
        >
          
            SORA
            
          <s-tooltip-stub
            animation="el-fade-in-linear"
            borderradius="small"
            closedelay="0"
            content="Copy Asset ID"
            hideafter="0"
            offset="0"
            opendelay="500"
            placement="top"
            showarrow="true"
            tabindex="-1"
            theme="auto"
            usedesignsystem="true"
          >
            <span
              class="asset-id"
            >
              
                (0x020...00000)
              
            </span>
          </s-tooltip-stub>
        </div>
      </div>
    </s-float-input-stub>
     
    <s-button-stub
      borderradius="small"
      class="wallet-send-action s-typography-button--large"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
      usedesignsystem="true"
    >
      
        Enter address
      
    </s-button-stub>
     
    <wallet-fee-stub
      value="0"
    />
  </div>
</wallet-base-stub>
`;

exports[`WalletSend.vue [Create Step: Incorrect Address]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="1"
  showback="true"
  showheader="true"
  title="Send"
  tooltip="Send tokens between SORA network accounts"
>
  <div
    class="wallet-send"
  >
    <address-book-input-stub
      class="wallet-send-address"
      excludeconnected="true"
      isvalid="true"
      value="12345"
    />
     
    <!---->
     
    <!---->
     
    <s-float-input-stub
      class="wallet-send-input"
      decimals="18"
      delimiters="[object Object]"
      haslocalestring="true"
      max="100000000000000000000"
      size="medium"
      value=""
    >
      <div
        class="wallet-send-amount"
      >
        <div
          class="wallet-send-amount-title"
        >
          Amount
        </div>
         
        <div
          class="wallet-send-amount-balance"
        >
          <span
            class="wallet-send-amount-balance-title"
          >
            Balance
          </span>
           
          <formatted-amount-with-fiat-value-stub
            assetsymbol="XOR"
            fiatfontsizerate="normal"
            fiatfontweightrate="normal"
            fiatformatasvalue="true"
            fiatvalue="151.782"
            fontsizerate="normal"
            fontweightrate="normal"
            hasfiatvalue="true"
            value="123.4"
            valuecanbehidden="true"
            valueclass="wallet-send-amount-balance-value"
            withleftshift="true"
          />
        </div>
      </div>
       
      <div
        class="asset s-flex"
      >
        <!---->
         
        <div
          class="asset-box"
        >
          <div
            class="asset-box__logo"
          >
            <token-logo-stub
              size="small"
              token="[object Object]"
              tokensymbol=""
            />
          </div>
           
          <span
            class="asset-name"
          >
            XOR
          </span>
        </div>
      </div>
       
      <div
        class="asset-info"
      >
        <formatted-amount-stub
          assetsymbol=""
          fiatsign="$"
          fontsizerate="normal"
          fontweightrate="normal"
          isfiatvalue="true"
          value="0"
        />
         
        <div
          class="asset-highlight"
        >
          
            SORA
            
          <s-tooltip-stub
            animation="el-fade-in-linear"
            borderradius="small"
            closedelay="0"
            content="Copy Asset ID"
            hideafter="0"
            offset="0"
            opendelay="500"
            placement="top"
            showarrow="true"
            tabindex="-1"
            theme="auto"
            usedesignsystem="true"
          >
            <span
              class="asset-id"
            >
              
                (0x020...00000)
              
            </span>
          </s-tooltip-stub>
        </div>
      </div>
    </s-float-input-stub>
     
    <s-button-stub
      borderradius="small"
      class="wallet-send-action s-typography-button--large"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
      usedesignsystem="true"
    >
      
        Enter amount
      
    </s-button-stub>
     
    <wallet-fee-stub
      value="0"
    />
  </div>
</wallet-base-stub>
`;

exports[`WalletSend.vue [Create Step: Not SORA Address]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="1"
  showback="true"
  showheader="true"
  title="Send"
  tooltip="Send tokens between SORA network accounts"
>
  <div
    class="wallet-send"
  >
    <address-book-input-stub
      class="wallet-send-address"
      excludeconnected="true"
      isvalid="true"
      value="5FbFx7AoLdnyMbKCbVZVZqXcZQkhy4HDKnerSW4oKeEQb6pL"
    />
     
    <p
      class="wallet-send-address-warning"
    >
      Tokens will be sent to the SORA address below. Make sure the recipient can access the SORA chain to use the tokens:
    </p>
     
    <s-tooltip-stub
      animation="el-fade-in-linear"
      borderradius="small"
      closedelay="0"
      content="Copy Asset ID"
      hideafter="0"
      offset="0"
      opendelay="500"
      placement="top"
      showarrow="true"
      tabindex="0"
      theme="auto"
      usedesignsystem="true"
    >
      <p
        class="wallet-send-address-formatted"
      >
        
            cnV1qZzFyBzkXS4xfmEnXue4SPhAbFNqcvULKSdzkfPEEZsBe
          
      </p>
    </s-tooltip-stub>
     
    <!---->
     
    <s-float-input-stub
      class="wallet-send-input"
      decimals="18"
      delimiters="[object Object]"
      haslocalestring="true"
      max="100000000000000000000"
      size="medium"
      value=""
    >
      <div
        class="wallet-send-amount"
      >
        <div
          class="wallet-send-amount-title"
        >
          Amount
        </div>
         
        <div
          class="wallet-send-amount-balance"
        >
          <span
            class="wallet-send-amount-balance-title"
          >
            Balance
          </span>
           
          <formatted-amount-with-fiat-value-stub
            assetsymbol="XOR"
            fiatfontsizerate="normal"
            fiatfontweightrate="normal"
            fiatformatasvalue="true"
            fiatvalue="151.782"
            fontsizerate="normal"
            fontweightrate="normal"
            hasfiatvalue="true"
            value="123.4"
            valuecanbehidden="true"
            valueclass="wallet-send-amount-balance-value"
            withleftshift="true"
          />
        </div>
      </div>
       
      <div
        class="asset s-flex"
      >
        <!---->
         
        <div
          class="asset-box"
        >
          <div
            class="asset-box__logo"
          >
            <token-logo-stub
              size="small"
              token="[object Object]"
              tokensymbol=""
            />
          </div>
           
          <span
            class="asset-name"
          >
            XOR
          </span>
        </div>
      </div>
       
      <div
        class="asset-info"
      >
        <formatted-amount-stub
          assetsymbol=""
          fiatsign="$"
          fontsizerate="normal"
          fontweightrate="normal"
          isfiatvalue="true"
          value="0"
        />
         
        <div
          class="asset-highlight"
        >
          
            SORA
            
          <s-tooltip-stub
            animation="el-fade-in-linear"
            borderradius="small"
            closedelay="0"
            content="Copy Asset ID"
            hideafter="0"
            offset="0"
            opendelay="500"
            placement="top"
            showarrow="true"
            tabindex="-1"
            theme="auto"
            usedesignsystem="true"
          >
            <span
              class="asset-id"
            >
              
                (0x020...00000)
              
            </span>
          </s-tooltip-stub>
        </div>
      </div>
    </s-float-input-stub>
     
    <s-button-stub
      borderradius="small"
      class="wallet-send-action s-typography-button--large"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
      usedesignsystem="true"
    >
      
        Enter amount
      
    </s-button-stub>
     
    <wallet-fee-stub
      value="0"
    />
  </div>
</wallet-base-stub>
`;

exports[`WalletSend.vue [Create Step: The Same Account]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="1"
  showback="true"
  showheader="true"
  title="Send"
  tooltip="Send tokens between SORA network accounts"
>
  <div
    class="wallet-send"
  >
    <address-book-input-stub
      class="wallet-send-address"
      excludeconnected="true"
      value="dfsakljfdlkjfhfkjladshslfjafds"
    />
     
    <!---->
     
    <p
      class="wallet-send-address-error"
    >
      You cannot send tokens to the same acount that you are currently connected to.
    </p>
     
    <s-float-input-stub
      class="wallet-send-input"
      decimals="18"
      delimiters="[object Object]"
      haslocalestring="true"
      max="100000000000000000000"
      size="medium"
      value=""
    >
      <div
        class="wallet-send-amount"
      >
        <div
          class="wallet-send-amount-title"
        >
          Amount
        </div>
         
        <div
          class="wallet-send-amount-balance"
        >
          <span
            class="wallet-send-amount-balance-title"
          >
            Balance
          </span>
           
          <formatted-amount-with-fiat-value-stub
            assetsymbol="XOR"
            fiatfontsizerate="normal"
            fiatfontweightrate="normal"
            fiatformatasvalue="true"
            fiatvalue="151.782"
            fontsizerate="normal"
            fontweightrate="normal"
            hasfiatvalue="true"
            value="123.4"
            valuecanbehidden="true"
            valueclass="wallet-send-amount-balance-value"
            withleftshift="true"
          />
        </div>
      </div>
       
      <div
        class="asset s-flex"
      >
        <!---->
         
        <div
          class="asset-box"
        >
          <div
            class="asset-box__logo"
          >
            <token-logo-stub
              size="small"
              token="[object Object]"
              tokensymbol=""
            />
          </div>
           
          <span
            class="asset-name"
          >
            XOR
          </span>
        </div>
      </div>
       
      <div
        class="asset-info"
      >
        <formatted-amount-stub
          assetsymbol=""
          fiatsign="$"
          fontsizerate="normal"
          fontweightrate="normal"
          isfiatvalue="true"
          value="0"
        />
         
        <div
          class="asset-highlight"
        >
          
            SORA
            
          <s-tooltip-stub
            animation="el-fade-in-linear"
            borderradius="small"
            closedelay="0"
            content="Copy Asset ID"
            hideafter="0"
            offset="0"
            opendelay="500"
            placement="top"
            showarrow="true"
            tabindex="-1"
            theme="auto"
            usedesignsystem="true"
          >
            <span
              class="asset-id"
            >
              
                (0x020...00000)
              
            </span>
          </s-tooltip-stub>
        </div>
      </div>
    </s-float-input-stub>
     
    <s-button-stub
      borderradius="small"
      class="wallet-send-action s-typography-button--large"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
      usedesignsystem="true"
    >
      
        Incorrect address
      
    </s-button-stub>
     
    <wallet-fee-stub
      value="0"
    />
  </div>
</wallet-base-stub>
`;
